pyvista/core/utilities/geometric_objects.py

```diff
@@ -363,6 +363,323 @@ def Sphere(
+
+    if theta[0] == 0.0 and theta[-1] == 2 * np.pi:
+        duplicate_theta = True
+        theta = theta[:-1]
+    else:
+        duplicate_theta = False
+
+    if phi[0] == 0.0:
+        points.extend(_spherical_to_cartesian(radius, phi[0], theta[0]))
+        positive_axis = True
+        phi = phi[1:]
+        nphi = nphi - 1
+        npoints_on_axis += nr
+    else:
+        positive_axis = False
+
+    if phi[-1] == np.pi:
+        points.extend(_spherical_to_cartesian(radius, phi[-1], theta[0]))
+        negative_axis = True
+        phi = phi[:-1]
+        nphi = nphi - 1
+        npoints_on_axis += nr
+    else:
+        negative_axis = False
+
+    # rest of points with theta changing quickest
+    for ir in radius:
+        for iphi in phi:
+            points.extend(_spherical_to_cartesian(ir, iphi, theta))
+
+    cells = []
+    celltypes = []
+
+    def _index(ir, iphi, itheta):
+        """Index for points not on axis.
+
+        Values of ir and phi here are relative to the first nonaxis values.
+        """
+        if duplicate_theta:
+            ntheta_ = ntheta - 1
+            itheta = itheta % ntheta_
+        else:
+            ntheta_ = ntheta
+
+        return npoints_on_axis + ir * nphi * ntheta_ + iphi * ntheta_ + itheta
+
+    if include_origin:
+        # first make the tetras that form with origin and axis point
+        #   origin is 0
+        #   first axis point is 1
+        #   other points at first phi position off axis
+        if positive_axis:
+            for itheta in range(ntheta - 1):
+                cells.append(4)
+                cells.extend([0, 1, _index(0, 0, itheta), _index(0, 0, itheta + 1)])
+
+                celltypes.append(pyvista.CellType.TETRA)
+
+        # next tetras that form with origin and bottom axis point
+        #   origin is 0
+        #   axis point is first in negative dir
+        #   other points at last phi position off axis
+        if negative_axis:
+            for itheta in range(ntheta - 1):
+                cells.append(4)
+                cells.extend(
+                    [
+                        0,
+                        npoints_on_axis_one_dir,
+                        _index(0, nphi - 1, itheta + 1),
+                        _index(0, nphi - 1, itheta),
+                    ]
+                )
+
+                celltypes.append(pyvista.CellType.TETRA)
+
+        # Pyramids that form to origin but without an axis point
+        for iphi in range(nphi - 1):
+            for itheta in range(ntheta - 1):
+                cells.append(5)
+                cells.extend(
+                    [
+                        _index(0, iphi, itheta),
+                        _index(0, iphi, itheta + 1),
+                        _index(0, iphi + 1, itheta + 1),
+                        _index(0, iphi + 1, itheta),
+                        0,
+                    ]
+                )
+
+                celltypes.append(pyvista.CellType.PYRAMID)
```

Comment:
To avoid deep nesting, it would be a good idea to use `itertools.product` in this case.
```suggestion
        for iphi, itheta in  itertools.product(range(nphi - 1), range(ntheta - 1)):
            cells.append(5)
            cells.extend(
                [
                    _index(0, iphi, itheta),
                    _index(0, iphi, itheta + 1),
                    _index(0, iphi + 1, itheta + 1),
                    _index(0, iphi + 1, itheta),
                    0,
                ]
            )

            celltypes.append(pyvista.CellType.PYRAMID)
```